core/xmodel/cross_query_cache.go

```diff
@@ -0,0 +1,269 @@
+	return &CrossQueryCache{
+		lg:          lg,
+		isPenetrate: true,
+	}
+}
+
+// NewCrossQueryCacheWirthData return CrossQuery instance while posttx
+func NewCrossQueryCacheWirthData(crossQueries []*pb.CrossQueryInfo) *CrossQueryCache {
+	lg := log.New("module", "cross_query_cache")
+	lg.SetHandler(log.StreamHandler(os.Stderr, log.LogfmtFormat()))
+	return &CrossQueryCache{
+		lg:               lg,
+		crossQueryCaches: crossQueries,
+		isPenetrate:      false,
+	}
+}
+
+// CrossQuery query contract from otherchain
+func (cqc *CrossQueryCache) CrossQuery(
+	crossQueryRequest *pb.CrossQueryRequest,
+	queryMeta *pb.CrossQueryMeta) (*pb.ContractResponse, error) {
+	cqc.lg.Info("Receive CrossQuery", "crossQueryRequest", crossQueryRequest, "queryMeta", queryMeta)
+	if !isQueryMetaValid(queryMeta) {
+		return nil, fmt.Errorf("isQueryParamValid check failed")
+	}
+	// Call endorsor for responce
+	if cqc.isPenetrate {
+		queryInfo, err := cqc.crossQueryFromEndorsor(crossQueryRequest, queryMeta)
+		if err != nil {
+			cqc.lg.Info("crossQueryFromEndorsor error", "error", err.Error())
+			return nil, err
+		}
+		cqc.crossQueryCaches = append(cqc.crossQueryCaches, queryInfo)
+		return queryInfo.GetResponse().GetResponse(), nil
+	}
+
+	// 验证背书规则、参数有效性、时间戳有效性
+	crossQuery := cqc.crossQueryCaches[cqc.crossQueryIdx]
+
+	// 验证request、签名等信息
+	for !cqc.isCossQueryValid(crossQueryRequest, queryMeta, crossQuery) {
+		return nil, fmt.Errorf("isCossQueryValid check failed")
+	}
+	cqc.crossQueryIdx++
+	return crossQuery.GetResponse().GetResponse(), nil
+}
+
+// isQueryParamValid 验证 query meta 背书策略是否有效
+func isQueryMetaValid(queryMeta *pb.CrossQueryMeta) bool {
+	return len(queryMeta.GetEndorsors()) >= int(queryMeta.GetChainMeta().GetMinEndorsorNum())
+}
+
+// crossQueryFromEndorsor will query cross from endorsor
+func (cqc *CrossQueryCache) crossQueryFromEndorsor(
+	crossQueryRequest *pb.CrossQueryRequest,
+	queryMeta *pb.CrossQueryMeta) (*pb.CrossQueryInfo, error) {
+
+	reqData, err := json.Marshal(crossQueryRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &pb.EndorserRequest{
+		RequestName: "CrossQueryPreExec",
+		BcName:      crossQueryRequest.GetBcname(),
+		RequestData: reqData,
+	}
+
+	res, signs, err := cqc.endorsorQueryWithGroup(req, queryMeta)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.CrossQueryInfo{
+		Request:  crossQueryRequest,
+		Response: res,
+		Signs:    signs,
+	}, nil
+}
+
+func (cqc *CrossQueryCache) endorsorQueryWithGroup(req *pb.EndorserRequest, queryMeta *pb.CrossQueryMeta) (*pb.CrossQueryResponse, []*pb.SignatureInfo, error) {
+	wg := sync.WaitGroup{}
+	msgChan := make(chan *queryRes, len(queryMeta.GetEndorsors()))
+
+	for idx := range queryMeta.GetEndorsors() {
+		wg.Add(1)
+		go func(req *pb.EndorserRequest, ce *pb.CrossEndorsor) {
+			defer wg.Done()
+			res, err := cqc.endorsorQuery(req, ce)
+			if err != nil {
+				return
+			}
+			msgChan <- res
+		}(req, queryMeta.GetEndorsors()[idx])
+	}
+	wg.Wait()
```

Comment:
Waiting for all endorser to return may take a long time, especially if `min_endorsor_num` is relatively small, not all the results are required. It is recommended to send the request while verifying the request, and if the number of signatures is sufficient, canel other requests directly.

Reply:
But we require that all endorsors's response equal. So cancel other requests directly may not good.